### DISCUSSION TITLE
sophacles/extra-widgets - StyledList and Macros.

### DISCUSSION THREAD

**Discussion by sophacles** [ratatui/ratatui#1439] [Ideas]
## Problem
I have a [repo]( https://github.com/sophacles/extra-widgets) of some widgets I've generalized from a couple personal projects.

Widgets:

SytledList - Similar to `widgets::List`, the `StyledList` is a list widget. It offers more
display/decoration options, optional separators between `ListItem`s and fixes a handful
of small annoyances (e.g. a selected multi-line item will always be fully displayed).

Calendar - Display a calendar for a month. This has some styling options, such as 
including day headers (M, T, W ...) and whether fill the grid with days from other months.
It also has the ability to style the individual dates.

(not actually a widget) text_macros - Some macros for reducing boilerplate when working
with larger text regions. For example, rather than:
```
Span::styled(
	"bold italic words",
	Style::default()
		.add_modifier(Modifier::BOLD)
		.add_modifier(Modifier::ITALIC),
);
```
one could write:

```
bold!(italic!("bold italic words"))
```

I'm not sure of the best way to share these.

## Solutions

I see 3 possible solutions to "how do I share these?":
* include them in the `widgets` module of ratatui. I'm happy to contribute them if there's interest.
* use them to seed a sibling widget collections (`ratatui-contrib` or similar). 
* release them as an independent crate (`extra-widgets` is a terrible name, I'm open to suggestions!)

There are pros and cons to each of the approaches. 

Having a project approved library of widgets is nice for users - there are more building blocks to make software from, and the widgets from the "blessed" library likely to be kept up-to-date with the ratatui core changes. This adds more work for maintainers, but also provides an easy mechanism to ensure that components don't lag behind ratatui itself (e.g. if a widget crate builds against ratatui version x, but a project builds against ratatui version y, and wants to include the widget, there may be compatibility issues).  

Having individual crates for widgets allows for more exploration and various ways to implement a concept, and and allows for more change in any given widget - a standard library of widgets, like a standard library for a programming language, may stagnate to maintain compatability for older programs and doing each widget as an independent crate helps avoid that. It hurts discoverability and crates run the risk of being abandoned forcing users to find a new fork (at best) or a new widget.

## Additional context
In the linked [repo](https://github.com/sophacles/extra-widgets) there are example programs for each widget in action. Additionally, in-depth rust docs can be generated from the repo - the `--document-private-items` flag provides even more info.

**Comment by sayanarijit** [comment]
Hi, we'll discuss this in a meeting. If you are interested in joining, pls mention your availability: https://calendly.com/d/y2h-b3c-n7s/ratatui-3

**Comment by joshka** [comment]
Neat! I wonder how much of your `List` changes could be brought into the main list widget?

**Comment by mindoodoo** [comment]
Hey, we'll be creating a discussion discussing the addition of macros.

We'd be interested in including the calendar widget inside the ratatui crate ! Would you be able to create a PR for it ?

**Comment by sophacles** [comment]
Sure thing, PR at #138

**Comment by joshka** [comment]
Doing some triage of old issues.
- Calendar ended up being included in the main repo.
- The macros would be a good fit moved into [ratatui-macros](https://github.com/ratatui-org/ratatui-macros)
- The styled list has some ideas that would be good to include in the core list. It would be good to list out the added functions and evaluate them

**Comment by orhun** [comment]
Btw I tried to reach out to @sophacles a couple of months ago but didn't hear back yet. I hope he's doing fine.

**Comment by joshka** [comment]
I'm going through the issue list to triage. Macro idea posted in ratatui-macros, calendar already dealt with, leaving only  the list to discuss here.

Here are the features that I see that it has over the built-in list. There's a bunch which would make good ideas for ratatui to incorporate.

1. Ability to separate each list item. Uses a line of half blocks (`"▄"`) to apply a fg/bg color to half of the line above / below each item (which changes based on the selection. This is a neat idea, and I wonder if it would make sense to extend this to rendering other characters as separators. Would be good to explore this
   <img width="139" alt="image" src="https://github.com/user-attachments/assets/3d557971-48c9-42f4-8c32-cb27d08ef428">
2. Highlight is renamed to selection for purposes of styling (this is reasonable, and something that would probably be a good idea for the main list to explore (assuming that we can migrate this in a non-breaking manner easily)
3. Highlight symbol is changed support both left and right symbols, which makes putting the characters at each end of the list item possible. I like this idea
  <img width="373" alt="image" src="https://github.com/user-attachments/assets/ced16924-1691-4b48-a27b-513c7ce91ee7"><br>
   This also supports configuring symbols that apply to every line, first line, last line, specific index / last line allowing for more configurability about how the symbols are applied. In doing so, this only supports a single character to be used / allocated space. I think this is an area to gain some ideas from.
   The same indicator config can be made on each item regardless of whether the item is selected. This is also a nice idea.
4. The list supports keeping a certain window of items visible (so that the item stays in the same location on the screen). Space is added above the list if there are not enough items in the window size:
   <img width="376" alt="image" src="https://github.com/user-attachments/assets/b1d8e3f0-3074-497e-9e21-f48b4df3648e"><br>
   again, I think this is a good idea. I think we had a recent PR that did something similar, but used a fixed number  for this rather than an enum for config (I like the enum approach better) https://docs.rs/ratatui/latest/ratatui/widgets/struct.List.html#method.scroll_padding
6. Overall the implementing code seems a bit neater than the ratatui implementation (particularly in the render function), so it's worth reading that and getting some ideas generally about how to render the list

On what to do with this, I see 3 possible options:
1. publish it as a list crate on its own. This would keep two lists around, meaning that we end up either duplicating effort to add ratatui features to this list, or the other way around.
2. Add the features to the built-in ratatui list.
3. Add the features to the tui-widget-list, which is proceeding on a faster iteration speed than the ratatui one (I'll link this comment there for their thoughts.)

If we were to take inspiration into Ratatui from this, that might look something like:
- [ ] Simplify List::render implementation looking at how this does it
- [ ] Rename highlight to selected
- [ ] Support left and right indicators for selected item instead of just the left item
- [ ] Configure how selection indicators work with an enum instead of just a single string
- [ ] Configure how windowing works with an enum instead of a single number
- [ ] Support separators and special rendering for separator lines

The big thing that's important is to make it easy for existing users to migrate to the new list (i.e. make these changes in a way that points them to the new implementations rather than just breaking apps)

I'd also consider whether it's worth moving the list out of ratatui core and into its own library `ratatui-list` that has a faster release cadence, to iterate on some of these things faster (and import that from the main library).
